go/vt/vtgate/planbuilder/testdata/reference_cases.json

```diff
@@ -0,0 +1,644 @@
+        "Variant": "Join"
+      },
+      "Original": "select user.col from user join main.ambiguous_ref_with_source",
+      "QueryType": "SELECT"
+    },
+    "gen4-plan": {
+      "Instructions": {
+        "FieldQuery": "select `user`.col from `user`, ambiguous_ref_with_source where 1 != 1",
+        "Keyspace": {
+          "Name": "user",
+          "Sharded": true
+        },
+        "OperatorType": "Route",
+        "Query": "select `user`.col from `user`, ambiguous_ref_with_source",
+        "Table": "`user`, ambiguous_ref_with_source",
+        "Variant": "Scatter"
+      },
+      "Original": "select user.col from user join main.ambiguous_ref_with_source",
+      "QueryType": "SELECT",
+      "TablesUsed": [
+        "main.ambiguous_ref_with_source",
+        "user.user"
+      ]
+    }
+  },
+  {
+    "comment": "insert into ambiguous qualified reference table routes to source",
+    "query": "insert into ambiguous_ref_with_source(col) values(1)",
+    "v3-plan": {
+      "Instructions": {
+        "Keyspace": {
+          "Name": "main",
+          "Sharded": false
+        },
+        "MultiShardAutocommit": false,
+        "OperatorType": "Insert",
+        "Query": "insert into ambiguous_ref_with_source(col) values (1)",
+        "TableName": "ambiguous_ref_with_source",
+        "TargetTabletType": "PRIMARY",
+        "Variant": "Unsharded"
+      },
+      "Original": "insert into ambiguous_ref_with_source(col) values(1)",
+      "QueryType": "INSERT",
+      "TablesUsed": [
+        "main.ambiguous_ref_with_source"
+      ]
+    },
+    "gen4-plan": {
+      "Instructions": {
+        "Keyspace": {
+          "Name": "main",
+          "Sharded": false
+        },
+        "MultiShardAutocommit": false,
+        "OperatorType": "Insert",
+        "Query": "insert into ambiguous_ref_with_source(col) values (1)",
+        "TableName": "ambiguous_ref_with_source",
+        "TargetTabletType": "PRIMARY",
+        "Variant": "Unsharded"
+      },
+      "Original": "insert into ambiguous_ref_with_source(col) values(1)",
+      "QueryType": "INSERT",
+      "TablesUsed": [
+        "main.ambiguous_ref_with_source"
+      ]
+    }
+  },
+  {
+    "comment": "insert into qualified ambiguous reference table routes requested keyspace",
+    "query": "insert into user.ambiguous_ref_with_source(col) values(1)",
+    "v3-plan": {
+      "Instructions": {
+        "Keyspace": {
+          "Name": "user",
+          "Sharded": true
+        },
+        "MultiShardAutocommit": false,
+        "OperatorType": "Insert",
+        "Query": "insert into ambiguous_ref_with_source(col) values (1)",
+        "TableName": "ambiguous_ref_with_source",
+        "TargetTabletType": "PRIMARY",
+        "Variant": "Sharded"
+      },
+      "Original": "insert into user.ambiguous_ref_with_source(col) values(1)",
+      "QueryType": "INSERT",
+      "TablesUsed": [
+        "user.ambiguous_ref_with_source"
+      ]
+    },
+    "gen4-plan": {
```

Comment:
I was expecting an error here because `UPDATE` and `DELETE` give errors. Added a similar (passing) test to `dml_cases.json` and tested it on `main`.

Reply:
true, I checked the execution will fail for that insert plan, ideally should fail at planning time.

Reply:
Note to sell, still TODO: route gen4 ref inserts back to source.

Reply:
Done